README.md
[![pypi](https://img.shields.io/pypi/v/vcfpy.svg)](https://pypi.python.org/pypi/vcfpy)
[![bioconda](https://img.shields.io/conda/dn/bioconda/vcfpy.svg?label=Bioconda)](https://bioconda.github.io/recipes/vcfpy/README.html)
[![CI](https://github.com/bihealth/vcfpy/actions/workflows/main.yml/badge.svg)](https://github.com/bihealth/vcfpy/actions/workflows/main.yml)
[![Documentation Status](https://readthedocs.org/projects/vcfpy/badge/?version=latest)](https://vcfpy.readthedocs.io/en/latest/?badge=latest)
[![Publication in The Journal of Open Source Software](http://joss.theoj.org/papers/edae85d90ea8a49843dbaaa109e47cba/status.svg)](http://joss.theoj.org/papers/10.21105/joss.00085)

# VCFPy

Python 3 VCF library with good support for both reading and writing

- Free software: MIT license
- Documentation: <https://vcfpy.readthedocs.io>.

## Features

- Support for reading and writing VCF v4.3
- Interface to `INFO` and `FORMAT` fields is based on `OrderedDict` allows for easier modification than PyVCF (also I find this more pythonic)
- Read (and jump in) and write BGZF files just using `vcfpy`

## Why another VCF parser for Python!

I've been using PyVCF with quite some success in the past. However, the
main bottleneck of PyVCF is when you want to modify the per-sample
genotype information. There are some issues in the tracker of PyVCF but
none of them can really be considered solved. I tried several hours to
solve these problems within PyVCF but this never got far or towards a
complete rewrite...

For this reason, VCFPy was born and here it is!

## What's the State?

VCFPy is the result of two full days of development plus some
maintenance work later now (right now). I'm using it in several projects
but it is not as battle-tested as PyVCF.

## Why Python 3 Only?

As I'm only using Python 3 code, I see no advantage in carrying around
support for legacy Python 2 and maintaining it. At a later point when
VCFPy is known to be stable, Python 2 support might be added if someone
contributes a pull request.



HISTORY.rst
=======
History
=======

v0.13.6 (2022-11-28)
--------------------

- Fixing bug in ``setup.py`` that prevented ``pysam`` dependency to be loaded (#150).

v0.13.5 (2022-11-13)
--------------------

- Treat ``.bgz`` files the same as ``.gz`` (#145, #149)

v0.13.4 (2022-04-13)
--------------------

- Switching to Github Actions for CI
- Fix INFO flag raises TypeError (#146)

v0.13.3 (2020-09-14)
--------------------

- Adding ``Record.update_calls``.
- Making ``Record.{format,calls}`` use list when empty

v0.13.2 (2020-08-20)
--------------------

- Adding ``Call.set_genotype()``.

v0.13.1 (2020-08-20)
--------------------

- Fixed ``Call.ploidy``.
- Fixed ``Call.is_variant``.

v0.13.0 (2020-07-10)
--------------------

* Fixing bug in case ``GT`` describes only one allele.
* Proper escaping of colon and semicolon (or the lack of escaping) in ``INFO`` and ``FORMAT``.

v0.12.2 (2020-04-29)
--------------------

* Fixing bug in case ``GT`` describes only one allele.

v0.12.1 (2019-03-08)
--------------------

* Not warning on ``PASS`` filter if not defined in header.

v0.12.0 (2019-01-29)
--------------------

* Fixing tests for Python >=3.6
* Fixing CI, improving tox integration.
* Applying ``black`` formatting.
* Replacing Makefile with more minimal one.
* Removing some linting errors from flake8.
* Adding support for reading VCF without ``FORMAT`` or any sample column.
* Adding support for writing headers and records without ``FORMAT`` and any sample columns.

v0.11.2 (2018-04-16)
--------------------

* Removing ``pip`` module from ``setup.py`` which is not recommended anyway.

v0.11.1 (2018-03-06)
--------------------

* Working around problem in HTSJDK output with incomplete ``FORMAT`` fields (#127).
  Writing out ``.`` instead of keeping trailing empty records empty.

v0.11.0 (2017-11-22)
--------------------

* The field ``FORMAT/FT`` is now expected to be a semicolon-separated string.
  Internally, we will handle it as a list.
* Switching from warning helper utility code to Python ``warnings`` module.
* Return ``str`` in case of problems with parsing value.

v0.10.0 (2017-02-27)
--------------------

* Extending API to allow for reading subsets of records.
  (Writing for sample subsets or reordered samples is possible through using the appropriate ``names`` list in the ``SamplesInfos`` for the ``Writer``).
* Deep-copying header lines and samples infos on ``Writer`` construction
* Using ``samples`` attribute from ``Header`` in ``Reader`` and ``Writer`` instead of passing explicitely

0.9.0 (2017-02-26)
------------------

* Restructuring of requirements.txt files
* Fixing parsing of no-call ``GT`` fields

0.8.1 (2017-02-08)
------------------

* PEP8 style adjustments
* Using versioneer for versioning
* Using ``requirements*.txt`` files now from setup.py
* Fixing dependency on cyordereddict to be for Python <3.6 instead of <3.5
* Jumping by samtools coordinate string now also allowed

0.8.0 (2016-10-31)
------------------

* Adding ``Header.has_header_line`` for querying existence of header line
* ``Header.add_*_line`` return a ``bool`` no indicating any conflicts
* Construction of Writer uses samples within header and no extra parameter (breaks API)

0.7.0 (2016-09-25)
------------------

* Smaller improvements and fixes to documentation
* Adding Codacy coverage and static code analysis results to README
* Various smaller code cleanup triggered by Codacy results
* Adding ``__eq__``, ``__neq__`` and ``__hash__`` to data types (where applicable)

0.6.0 (2016-09-25
-----------------

* Refining implementation for breakend and symbolic allele class
* Removing ``record.SV_CODES``
* Refactoring parser module a bit to make the code cleaner
* Fixing small typos and problems in documentation

0.5.0 (2016-09-24)
------------------

* Deactivating warnings on record parsing by default because of performance
* Adding validation for ``INFO`` and ``FORMAT`` fields on reading (#8)
* Adding predefined ``INFO`` and ``FORMAT`` fields to ``pyvcf.header`` (#32)

0.4.1 (2016-09-22)
------------------

* Initially enabling codeclimate

0.4.0 (2016-09-22)
------------------

* Exporting constants for encoding variant types
* Exporting genotype constants ``HOM_REF``, ``HOM_ALT``, ``HET``
* Implementing ``Call.is_phased``, ``Call.is_het``, ``Call.is_variant``, ``Call.is_phased``, ``Call.is_hom_ref``, ``Call.is_hom_alt``
* Removing ``Call.phased`` (breaks API, next release is 0.4.0)
* Adding tests, fixing bugs for methods of ``Call``

0.3.1 (2016-09-21)
------------------

* Work around ``FORMAT/FT`` being a string; this is done so in the Delly output

0.3.0 (2016-09-21)
------------------

* ``Reader`` and ``Writer`` can now be used as context manager (with ``with``)
* Including license in documentation, including Biopython license
* Adding support for writing bgzf files (taken from Biopython)
* Adding support for parsing arrays in header lines
* Removing ``example-4.1-bnd.vcf`` example file because v4.1 tumor derival lacks ``ID`` field
* Adding ``AltAlleleHeaderLine``, ``MetaHeaderLine``, ``PedigreeHeaderLine``, and ``SampleHeaderLine``
* Renaming ``SimpleHeaderFile`` to ``SimpleHeaderLine``
* Warn on missing ``FILTER`` entries on parsing
* Reordered parameters in ``from_stream`` and ``from_file`` (#18)
* Renamed ``from_file`` to ``from_stream`` (#18)
* Renamed ``Reader.jump_to`` to ``Reader.fetch``
* Adding ``header_without_lines`` function
* Generally extending API to make it esier to use
* Upgrading dependencies, enabling pyup-bot
* Greatly extending documentation

0.2.1 (2016-09-19)
------------------

* First release on PyPI


vcfpy/record.py
# -*- coding: utf-8 -*-
"""Code for representing a VCF record

The VCF record structure is modeled after the one of PyVCF
"""

import re


#: Code for single nucleotide variant allele
SNV = "SNV"
#: Code for a multi nucleotide variant allele
MNV = "MNV"
#: Code for "clean" deletion allele
DEL = "DEL"
#: Code for "clean" insertion allele
INS = "INS"
#: Code for indel allele, includes substitutions of unequal length
INDEL = "INDEL"
#: Code for structural variant allele
SV = "SV"
#: Code for break-end allele
BND = "BND"
#: Code for symbolic allele that is neither SV nor BND
SYMBOLIC = "SYMBOLIC"

#: Code for mixed variant type
MIXED = "MIXED"

#: Code for homozygous reference
HOM_REF = 0
#: Code for heterozygous
HET = 1
#: Code for homozygous alternative
HOM_ALT = 2

#: Characters reserved in VCF, have to be escaped in INFO fields
RESERVED_CHARS = {"INFO": ";%,\r\n\t", "FORMAT": ":=%,\r\n\t"}
#: Mapping for escaping reserved characters
ESCAPE_MAPPING = [
    ("%", "%25"),
    (":", "%3A"),
    (";", "%3B"),
    ("=", "%3D"),
    (",", "%2C"),
    ("\r", "%0D"),
    ("\n", "%0A"),
    ("\t", "%09"),
]
#: Mapping from escaped characters to reserved one
UNESCAPE_MAPPING = [(v, k) for k, v in ESCAPE_MAPPING]


class Record:
    """Represent one record from the VCF file

    Record objects are iterators of their calls
    """

    def __init__(self, CHROM, POS, ID, REF, ALT, QUAL, FILTER, INFO, FORMAT=None, calls=None):
        if bool(FORMAT) != bool(calls):
            raise ValueError("Either provide both FORMAT and calls or none.")
        #: A ``str`` with the chromosome name
        self.CHROM = CHROM
        #: An ``int`` with a 1-based begin position
        self.POS = POS
        #: An ``int`` with a 0-based begin position
        self.begin = POS - 1
        #: An ``int`` with a 0-based end position
        self.end = None  # XXX
        #: A list of the semicolon-separated values of the ID column
        self.ID = list(ID)
        #: A ``str`` with the REF value
        self.REF = REF
        #: A list of alternative allele records of type :py:class:`AltRecord`
        self.ALT = list(ALT)
        #: The quality value, can be ``None``
        self.QUAL = QUAL
        #: A list of strings for the FILTER column
        self.FILTER = FILTER
        #: An OrderedDict giving the values of the INFO column, flags are
        #: mapped to ``True``
        self.INFO = INFO
        #: A list of strings for the FORMAT column.  Optional, must be given if
        #: and only if ``calls`` is also given.
        self.FORMAT = FORMAT or []
        #: A list of genotype :py:class:`Call` objects.  Optional, must be given if
        #: and only if ``FORMAT`` is also given.
        self.calls = calls or []
        #: A mapping from sample name to entry in self.calls.
        self.call_for_sample = {}
        self.update_calls(self.calls)

    def update_calls(self, calls):
        """Update ``self.calls`` and other fields as necessary."""
        for call in calls:
            call.site = self
        self.call_for_sample = {call.sample: call for call in calls}

    def is_snv(self):
        """Return ``True`` if it is a SNV"""
        return len(self.REF) == 1 and all(a.type == "SNV" for a in self.ALT)

    @property
    def affected_start(self):
        """Return affected start position in 0-based coordinates

        For SNVs, MNVs, and deletions, the behaviour is the start position.
        In the case of insertions, the position behind the insert position is
        returned, yielding a 0-length interval together with
        :py:meth:`~Record.affected_end`
        """
        types = {alt.type for alt in self.ALT}  # set!
        BAD_MIX = {INS, SV, BND, SYMBOLIC}  # don't mix well with others
        if (BAD_MIX & types) and len(types) == 1 and list(types)[0] == INS:
            # Only insertions, return 0-based position right of first base
            return self.POS  # right of first base
        else:  # Return 0-based start position of first REF base
            return self.POS - 1  # left of first base

    @property
    def affected_end(self):
        """Return affected start position in 0-based coordinates

        For SNVs, MNVs, and deletions, the behaviour is based on the start
        position and the length of the REF.  In the case of insertions, the
        position behind the insert position is returned, yielding a 0-length
        interval together with :py:meth:`~Record.affected_start`
        """
        types = {alt.type for alt in self.ALT}  # set!
        BAD_MIX = {INS, SV, BND, SYMBOLIC}  # don't mix well with others
        if (BAD_MIX & types) and len(types) == 1 and list(types)[0] == INS:
            # Only insertions, return 0-based position right of first base
            return self.POS  # right of first base
        else:  # Return 0-based end position, behind last REF base
            return (self.POS - 1) + len(self.REF)

    def add_filter(self, label):
        """Add label to FILTER if not set yet, removing ``PASS`` entry if
        present
        """
        if label not in self.FILTER:
            if "PASS" in self.FILTER:
                self.FILTER = [f for f in self.FILTER if f != "PASS"]
            self.FILTER.append(label)

    def add_format(self, key, value=None):
        """Add an entry to format

        The record's calls ``data[key]`` will be set to ``value`` if not yet
        set and value is not ``None``.  If key is already in FORMAT then
        nothing is done.
        """
        if key in self.FORMAT:
            return
        self.FORMAT.append(key)
        if value is not None:
            for call in self:
                call.data.setdefault(key, value)

    def __iter__(self):
        """Return generator yielding from ``self.calls``"""
        yield from self.calls

    def __eq__(self, other):
        if isinstance(other, self.__class__):
            return self.__dict__ == other.__dict__
        return NotImplemented

    def __ne__(self, other):
        if isinstance(other, self.__class__):
            return not self.__eq__(other)
        return NotImplemented

    def __hash__(self):
        return hash(tuple(sorted(self.__dict__.items())))

    def __str__(self):
        tpl = "Record({})"
        lst = [
            self.CHROM,
            self.POS,
            self.ID,
            self.REF,
            self.ALT,
            self.QUAL,
            self.FILTER,
            self.INFO,
            self.FORMAT,
            self.calls,
        ]
        return tpl.format(", ".join(map(repr, lst)))

    def __repr__(self):
        return str(self)


class UnparsedCall:
    """Placeholder for :py:class:`Call` when parsing only a subset of fields"""

    def __init__(self, sample, unparsed_data, site=None):
        #: the name of the sample for which the call was made
        self.sample = sample
        #: ``str`` with the unparsed data
        self.unparsed_data = unparsed_data
        #: the :py:class:`Record` of this :py:class:`Call`
        self.site = site


#: Regular expression for splitting alleles
ALLELE_DELIM = re.compile(r"[|/]")


class Call:
    """The information for a genotype callable

    By VCF, this should always include the genotype information and
    can contain an arbitrary number of further annotation, e.g., the
    coverage at the variant position.
    """

    def __init__(self, sample, data, site=None):
        #: the name of the sample for which the call was made
        self.sample = sample
        #: an OrderedDict with the key/value pair information from the
        #: call's data
        self.data = data
        #: the :py:class:`Record` of this :py:class:`Call`
        self.site = site
        #: the allele numbers (0, 1, ...) in this calls or None for no-call
        self.gt_alleles = None
        #: whether or not the variant is fully called
        self.called = None
        #: the number of alleles in this sample's call
        self.ploidy = None
        self._genotype_updated()

    def set_genotype(self, genotype):
        """Set ``self.data["GT"]`` to ``genotype`` and properly update related
        properties.
        """
        self.data["GT"] = genotype
        self._genotype_updated()

    def _genotype_updated(self):
        """Update fields related to ``self.data["GT"]``."""
        if self.data.get("GT", None) is None:
            self.gt_alleles = None
            self.called = None
            self.ploidy = None
        else:
            self.gt_alleles = []
            for allele in ALLELE_DELIM.split(str(self.data["GT"])):
                if allele == ".":
                    self.gt_alleles.append(None)
                else:
                    self.gt_alleles.append(int(allele))
            self.called = all([al is not None for al in self.gt_alleles])
            self.ploidy = len(self.gt_alleles)

    @property
    def is_phased(self):
        """Return boolean indicating whether this call is phased"""
        return "|" in self.data.get("GT", "")

    @property
    def gt_phase_char(self):
        """Return character to use for phasing"""
        return "/" if not self.is_phased else "|"

    @property
    def gt_bases(self):
        """Return the actual genotype bases, e.g. if VCF genotype is 0/1,
        could return ('A', 'T')
        """
        result = []
        for a in self.gt_alleles:
            if a is None:
                result.append(None)
            elif a == 0:
                result.append(self.site.REF)
            else:
                result.append(self.site.ALT[a - 1].value)
        return tuple(result)

    @property
    def gt_type(self):
        """The type of genotype, returns one of ``HOM_REF``, ``HOM_ALT``, and
        ``HET``.
        """
        if not self.called:
            return None  # not called
        elif all(a == 0 for a in self.gt_alleles):
            return HOM_REF
        elif len(set(self.gt_alleles)) == 1:
            return HOM_ALT
        else:
            return HET

    def is_filtered(self, require=None, ignore=None):
        """Return ``True`` for filtered calls

        :param iterable ignore: if set, the filters to ignore, make sure to
            include 'PASS', when setting, default is ``['PASS']``
        :param iterable require: if set, the filters to require for returning
            ``True``
        """
        ignore = ignore or ["PASS"]
        if "FT" not in self.data or not self.data["FT"]:
            return False
        for ft in self.data["FT"]:
            if ft in ignore:
                continue  # skip
            if not require:
                return True
            elif ft in require:
                return True
        return False

    @property
    def is_het(self):
        """Return ``True`` for heterozygous calls"""
        return self.gt_type == HET

    @property
    def is_variant(self):
        """Return ``True`` for non-hom-ref calls"""
        return bool(self.gt_type)

    def __eq__(self, other):
        if isinstance(other, self.__class__):
            return self.__dict__ == other.__dict__
        return NotImplemented

    def __ne__(self, other):
        if isinstance(other, self.__class__):
            return not self.__eq__(other)
        return NotImplemented

    def __hash__(self):
        return hash(tuple(sorted(self.__dict__.items())))

    def __str__(self):
        tpl = "Call({})"
        lst = [self.sample, self.data]
        return tpl.format(", ".join(map(repr, lst)))

    def __repr__(self):
        return str(self)


class AltRecord:
    """An alternative allele Record

    Currently, can be a substitution, an SV placeholder, or breakend
    """

    def __init__(self, type_=None):
        #: String describing the type of the variant, could be one of
        #: SNV, MNV, could be any of teh types described in the ALT
        #: header lines, such as DUP, DEL, INS, ...
        self.type = type_

    def __eq__(self, other):
        if isinstance(other, self.__class__):
            return self.__dict__ == other.__dict__
        return NotImplemented

    def __ne__(self, other):
        if isinstance(other, self.__class__):
            return not self.__eq__(other)
        return NotImplemented

    def __hash__(self):
        return hash(tuple(sorted(self.__dict__.items())))

    def serialize(self):
        """Return ``str`` with representation for VCF file"""
        raise NotImplementedError("Abstract class, implemented in sub class")


class Substitution(AltRecord):
    """A basic alternative allele record describing a REF->AltRecord
    substitution

    Note that this subsumes MNVs, insertions, and deletions.
    """

    def __init__(self, type_, value):
        super().__init__(type_)
        #: The alternative base sequence to use in the substitution
        self.value = value

    def serialize(self):
        return self.value

    def __eq__(self, other):
        if isinstance(other, self.__class__):
            return self.__dict__ == other.__dict__
        return NotImplemented

    def __ne__(self, other):
        if isinstance(other, self.__class__):
            return not self.__eq__(other)
        return NotImplemented

    def __hash__(self):
        return hash(tuple(sorted(self.__dict__.items())))

    def __str__(self):
        tpl = "Substitution(type_={}, value={})"
        return tpl.format(*map(repr, [self.type, self.value]))

    def __repr__(self):
        return str(self)


#: code for five prime orientation :py:class:`BreakEnd`
FIVE_PRIME = "5"
#: code for three prime orientation :py:class:`BreakEnd`
THREE_PRIME = "3"

#: code for forward orientation
FORWARD = "+"
#: code for reverse orientation
REVERSE = "-"


class BreakEnd(AltRecord):
    """A placeholder for a breakend"""

    def __init__(
        self, mate_chrom, mate_pos, orientation, mate_orientation, sequence, within_main_assembly
    ):
        super().__init__("BND")
        #: chromosome of the mate breakend
        self.mate_chrom = mate_chrom
        #: position of the mate breakend
        self.mate_pos = mate_pos
        #: orientation of this breakend
        self.orientation = orientation
        #: orientation breakend's mate
        self.mate_orientation = mate_orientation
        #: breakpoint's connecting sequence
        self.sequence = sequence
        #: ``bool`` specifying if the breakend mate is within the assembly
        #: (``True``) or in an ancillary assembly (``False``)
        self.within_main_assembly = within_main_assembly

    def serialize(self):
        """Return string representation for VCF"""
        if self.mate_chrom is None:
            remote_tag = "."
        else:
            if self.within_main_assembly:
                mate_chrom = self.mate_chrom
            else:
                mate_chrom = "<{}>".format(self.mate_chrom)
            tpl = {FORWARD: "[{}:{}[", REVERSE: "]{}:{}]"}[self.mate_orientation]
            remote_tag = tpl.format(mate_chrom, self.mate_pos)
        if self.orientation == FORWARD:
            return remote_tag + self.sequence
        else:
            return self.sequence + remote_tag

    def __eq__(self, other):
        if isinstance(other, self.__class__):
            return self.__dict__ == other.__dict__
        return NotImplemented

    def __ne__(self, other):
        if isinstance(other, self.__class__):
            return not self.__eq__(other)
        return NotImplemented

    def __hash__(self):
        return hash(tuple(sorted(self.__dict__.items())))

    def __str__(self):
        tpl = "BreakEnd({})"
        vals = [
            self.mate_chrom,
            self.mate_pos,
            self.orientation,
            self.mate_orientation,
            self.sequence,
            self.within_main_assembly,
        ]
        return tpl.format(", ".join(map(repr, vals)))

    def __repr__(self):
        return str(self)


class SingleBreakEnd(BreakEnd):
    """A placeholder for a single breakend"""

    def __init__(self, orientation, sequence):
        super().__init__(None, None, orientation, None, sequence, None)

    def __eq__(self, other):
        if isinstance(other, self.__class__):
            return self.__dict__ == other.__dict__
        return NotImplemented

    def __ne__(self, other):
        if isinstance(other, self.__class__):
            return not self.__eq__(other)
        return NotImplemented

    def __hash__(self):
        return hash(tuple(sorted(self.__dict__.items())))

    def __str__(self):
        tpl = "SingleBreakEnd({})"
        vals = [self.orientation, self.sequence]
        return tpl.format(", ".join(map(repr, vals)))


class SymbolicAllele(AltRecord):
    """A placeholder for a symbolic allele

    The allele symbol must be defined in the header using an ``ALT`` header
    before being parsed.  Usually, this is used for succinct descriptions of
    structural variants or IUPAC parameters.
    """

    def __init__(self, value):
        super().__init__(SYMBOLIC)
        #: The symbolic value, e.g. 'DUP'
        self.value = value

    def serialize(self):
        return "<{}>".format(self.value)

    def __eq__(self, other):
        if isinstance(other, self.__class__):
            return self.__dict__ == other.__dict__
        return NotImplemented

    def __ne__(self, other):
        if isinstance(other, self.__class__):
            return not self.__eq__(other)
        return NotImplemented

    def __hash__(self):
        return hash(tuple(sorted(self.__dict__.items())))

    def __str__(self):
        return "SymbolicAllele({})".format(repr(self.value))

    def __repr__(self):
        return str(self)
